src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/IRequestHandler.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+#nullable enable
+
+namespace CommonLanguageServerProtocol.Framework;
+
+/// <summary>
+/// Top level type for LSP request handler.
+/// </summary>
+public interface IRequestHandler<RequestContextType>
+{
+    /// <summary>
+    /// Whether or not the solution state on the server is modified
+    /// as a part of handling this request.
+    /// </summary>
+    bool MutatesSolutionState { get; }
+
+    /// <summary>
+    /// Whether or not the handler execution queue should build a solution that represents the LSP
+    /// state of the world. If this property is not set <see cref="RequestContext.Solution"/> will be <see langword="null"/>
+    /// and <see cref="RequestContext.Document"/> will be <see langword="null"/>, even if <see cref="IRequestHandler{RequestType, ResponseType, RequestContextType}.GetTextDocumentIdentifier(RequestType)"/>
+    /// doesn't return null. Handlers should still provide text document information if possible to
+    /// ensure the correct workspace is found and validated.
+    /// </summary>
+    bool RequiresLSPSolution { get; }
```

Comment:
Is it expensive to always provide the solution? I imagine we'd have it cached from the previuos mutation anyways?

Reply:
@dibarbet can you remind me what this item accomplishes? I agreed when you explained it to me but I've forgotten why now.